cpp/src/arrow/util/hashing.h

```diff
@@ -0,0 +1,805 @@
+      *out_data++ = cast_offset;
+    }
+  }
+
+  template <class Offset>
+  void CopyOffsets(Offset* out_data) const {
+    CopyOffsets(0, out_data);
+  }
+
+  // Copy values starting from index `start` into `out_data`
+  void CopyValues(int32_t start, uint8_t* out_data) const {
+    CopyValues(start, -1, out_data);
+  }
+
+  // Same as above, but check output size in debug mode
+  void CopyValues(int32_t start, int64_t out_size, uint8_t* out_data) const {
+    int32_t offset = offsets_[start];
+    auto length = values_.size() - static_cast<size_t>(offset);
+    if (out_size != -1) {
+      assert(static_cast<int64_t>(length) == out_size);
+    }
+    memcpy(out_data, values_.data() + offset, length);
+  }
+
+  void CopyValues(uint8_t* out_data) const { CopyValues(0, -1, out_data); }
+
+  void CopyValues(int64_t out_size, uint8_t* out_data) const {
+    CopyValues(0, out_size, out_data);
+  }
+
+ protected:
+  struct Payload {
+    int32_t memo_index;
+  };
+  using HashTableType = HashTable<Payload>;
+  using HashTableEntry = typename HashTable<Payload>::Entry;
+  HashTableType hash_table_;
+  std::vector<int32_t> offsets_;
+  std::string values_;
+
+  std::pair<const HashTableEntry*, bool> Lookup(hash_t h, const void* data,
+                                                int32_t length) const {
+    auto cmp_func = [=](const Payload* payload) {
+      int32_t start, stop;
+      start = offsets_[payload->memo_index];
+      stop = offsets_[payload->memo_index + 1];
+      return length == stop - start && memcmp(data, values_.data() + start, length) == 0;
+    };
+    return hash_table_.Lookup(h, cmp_func);
+  }
+};
+
+template <typename T, typename Enable = void>
+struct HashTraits {};
+
+template <>
+struct HashTraits<BooleanType> {
+  using MemoTableType = SmallScalarMemoTable<bool>;
+};
+
+template <typename T>
+struct HashTraits<T, enable_if_8bit_int<T>> {
+  using c_type = typename T::c_type;
+  using MemoTableType = SmallScalarMemoTable<typename T::c_type>;
+
+  static Status GetDictionaryArrayData(MemoryPool* pool,
+                                       const std::shared_ptr<DataType>& type,
+                                       const MemoTableType& memo_table,
+                                       int64_t start_offset,
+                                       std::shared_ptr<ArrayData>* out) {
+    std::shared_ptr<Buffer> dict_buffer;
+    auto dict_length = static_cast<int64_t>(memo_table.size()) - start_offset;
+    // This makes a copy, but we assume a dictionary array is usually small
+    // compared to the size of the dictionary-using array.
+    // (also, copying the dictionary values is cheap compared to the cost
+    //  of building the memo table)
+    RETURN_NOT_OK(
+        AllocateBuffer(pool, TypeTraits<T>::bytes_required(dict_length), &dict_buffer));
+    memo_table.CopyValues(static_cast<int32_t>(start_offset),
+                          reinterpret_cast<c_type*>(dict_buffer->mutable_data()));
```

Comment:
See my comments about this in `BinaryMemoTable`. I think we should try to avoid memory doubling here -- the memory doubling issue is more significant than the copying performance (which as you point out is not significant relative to the cost of building the table). It is true that in many cases the dictionary is small, but that won't stop people from creating multi-gigabyte dictionaries (and they do)